### DISCUSSION TITLE
Next.js | How many serverless functions actually created?

### DISCUSSION THREAD

**Discussion by RazerMoon** [vercel/vercel#5458] [Help] [answered]
Hi, I have been trying to figure out how many serverless functions actually get generated when using Next.js (so I can calculate if I won't exceed the limit) but the explanations sometimes contradict each other.

[This page](https://vercel.com/docs/platform/limits) states:

> NOTE: Next.js bundles all API Routes and server-rendered pages into **individual** Serverless Functions when deployed on Vercel. As a result, the "Serverless Functions per Deployment" limit is unlikely to apply for Next.js projects.

Which I interpret as:

- Page 1 with getServerSideProps gets a function
- Page 2 with getServerSideProps gets another
- API route 1 gets a third
- API route 2 gets a fourth

_Therefore_, subtract **four** from the monthly limit

---

However, a reply to a different discussion
_Originally posted by @samsisle in https://github.com/vercel/vercel/discussions/5093#discussioncomment-57628_ states:

> in Vercel (a recent update), Next.js SSR pages and pages/api routes get bundled into **two** separate Lambdas (λ), so you should not have to worry about the Serverless Function limit.

Which I interpret as:

- Page 1 with getServerSideProps gets a function
- Page 2 with getServerSideProps is bundled with previous
- API route 1 gets a second
- API route 2 is bundled with previous

_Therefore_, subtract **two** from the monthly limit

---

I would appreciate if someone gave me a breakdown, e.g. if I have: 

1. Two API routes
2. Two pages with getStaticProps
3. Two pages with getServerSideProps
4. Two pages with getStaticProps **that use ISR**
5. Page that uses getStaticProps **and** getStaticPaths

**How many do I subtract from the monthly limit?**

My current understanding:
1. Two API routes = one (they get bundled)
2. Two pages with getStaticProps = none (Static)
3. Two pages with getServerSideProps = one (they get bundled)
4. Two pages with getStaticProps **that use ISR** = two (function for each)
5. Page that uses getStaticProps **and** getStaticPaths = none (both are static)

_Therefore_, subtract **four** from the monthly limit.

**Comment by mcsdevv** [comment, marked as answer]
Hello, Next.js bundles functions, as a result, this limit rarely applies to such projects. In all your examples, this would equate to a single Serverless Function. The only time more would be created is when the size exceeds 50mb.

**Reply by RazerMoon** [reply]
Oh wow, looks like I was totally off the mark then 😅. Thank you for the reply.

**Reply by redbar0n** [reply]
@mcsdevv so you would still use 1 from the monthly limit?

**Reply by mcsdevv** [reply]
Almost certainly @redbar0n, in very rare cases I've seen 2.

**Reply by redbar0n** [reply]
@mcsdevv Thanks. I read from [someone who had been in contact with support](https://medium.com/@traderjosh16/i-emailed-vercel-support-about-this-and-was-told-they-recently-made-an-update-where-all-next-js-fada1fb4af02), and they had gotten in response that pages would take 1 function and API routes would take 1 function, so 2 in total. But that doesn't seem like a very rare case?

The [Limits page](https://vercel.com/docs/platform/limits) was very confusing:

>"Next.js bundles all API Routes and server-rendered pages into shared Serverless Functions, separating API routes and server-rendered pages when deployed on Vercel."

Bundled and separated at the same time?

**Reply by martinnabhan** [reply]
Hello @mcsdevv and thank you for all your informative responses.

We're currently prototyping a new project on Vercel using Next.js, and we've started to hit the Serverless function deployment limit. As far as I can tell from your responses above API Routes and SSR routes should be bundled and it should be near impossible to hit the deployment limit while using Next.js.

We've got 3 API Routes, with no plans on adding more. However the problem seems to be our SSR routes. We're using getServerSideProps in about 30 pages right now, with plans to add even more. It's my understanding that these would be bundled into 1(?) Serverless function, but this doesn't seem to be happening.

Could you please provide some more details of how the bundling works? Is there any way to see how many functions are being deployed after bundling?

We would love to use Vercel for our upcoming projects but as it stands right now we're going to be forced to use another provider, which I would really prefer not to, since I really love using Vercel.

Thanks!

**Reply by mcsdevv** [reply]
Hi @martinnabhan, API Routes will all be bundled, each SSR page will also be bundled - both of these will be creating another function should they exceed maximum capacity (50MB) although this is handled a little before the limit to avoid it being too close.

What plan are you on? Any commercial project won't have to worry about this at all.

The easiest way to see the build outputs is on the Vercel dashboard by selecting the individual deployments and selecting the 'Source' tab.

**Reply by martinnabhan** [reply]
Wow, thank you for the super swift reply!

We're on the Hobby plan now, planning to upgrade to the Pro plan if Vercel turns out to be viable.
Would the limit be increased by upgrading to the Pro plan? The Limits page currently states 12 deployments for the Hobby plan, while Pro and Enterprise are "-", which I'm not sure if it means "limitless" or "same as Hobby".

As for `getServerSideProps`, I also first thought each use of `getServerSideProps` would count as 1 against the limit,
but I've been trying out different deployments with less routes, and right now we're able to deploy 30(!) pages using getServerSideProps.

I'm not sure where the cut-off is, but there does seem to be some bundling going on, even with pages using `getServerSideProps`. If you could please come back to me when you know more I would be super thankful!

PS. While reading other threads on here I get the impression Vercel might not be made for lots of SSR routes. If that's the case I'll be sure to use Vercel for mainly ISG and SSG-heavy projects!

**Reply by mcsdevv** [reply]
Hi @martinnabhan, the Hobby plan is the only one limited in this aspect, neither the Pro or Enterprise plans are limited.

I've updated my previous comment to reflect my discussion with the Next.js team. The short version is you'd need a lot of extremely large dependencies in your project to hit the 12 limits and if it's on a Pro plan, there is no limit.

Vercel is absolutely a great place for all types of projects, SSG, SSR, and ISR - it's about guiding people to the right solution which very often SSG, and also sometimes SSR/ISR, dependent on their needs.

**Reply by martinnabhan** [reply]
Thank you so much for clarifying! You've been incredibly helpful.
We'll be upgrading promptly then!
I've had an amazing experience with Vercel so far and I'm super excited to continue building with both Vercel and Next.js.

Once again, thank you!

**Comment by eric-burel** [comment]
Hi everybody, is this answer still up to date in 2023? Namely, I wonder about the bundling process: [the documentation](https://vercel.com/docs/concepts/functions/serverless-functions) states that "in order to optimize resources, there is an internal process that bundles as many Serverless Functions as possible into a single Lambda" and that "the remaining functions will be bundled together optimizing for how many functions are created"

- Should I understand that functions may be bundled together depending on some usage statistics? For instance, the functions might be rebundled differently over time depending on the volumes of requests? Say I have a "/login" and a "/signup" page, if login is accessed more often than signup, can it end up into its own lambda?
- Conversely, if functions end up in the same lambda, do they share the same runtime? For instance I am using node-cache to reduce the database workload and I wonder how many different cache will exist, I guess at most one cache per page and API routes, but with bundling it may be much less
- Bonus: do you have any comparison with other hosting services? Will they bundle too or will they use 1 lambda per function?

**Reply by amyegan** [reply]
Hi @eric-burel! Bundling isn't my area of expertise, so I don't know the answer off the top of my head. But I'm going to do some digging and get the right info to update the docs. Just wanted to let you know someone saw this and is working on it!

**Reply by pirtlj** [reply]
I am also very interested in this.  

I am currently using auth0-next which has api callbacks for logging in, unfortunately these callbacks can't be migrated to edge functions so they get bungled into the lambda with all the other api routes. 

As the API gets bigger the cold starts get longer and this degrades the UX when a user wants to login and has to wait for a huge lambda to cold startup.  It would be great if we could control the code splitting of API routes into their own bundles. 

My current plan is to put the auth callbacks into a completely separate Next.js app and run it on a sub domain so they get their own lightweight lambda and populate the user session.

**Reply by eric-burel** [reply]
@pirtlj does this section help? https://vercel.com/docs/concepts/functions/serverless-functions#advanced-usage
I think you can configure the bundling by hand if you have API routes you want to live in their own lambdas, what's more opaque to me is how Vercel behave when bundling automatically when there is no config

**Reply by pirtlj** [reply]
Oh wow thanks!   I’ll try this out.  I didn’t realize you could have that granular of control.
